### ISSUE TITLE
%%javascript cells executing unconditionally for a trusted notebook

### ISSUE THREAD

**Issue by sjdv1982** [jupyter/notebook#3149]
Upon experimenting with %%javascript cells in notebooks, I was surprised to discover that they execute unconditionally upon loading a trusted notebook (even with F5).

This is a very pleasant surprise to me, please don't change this behavior!
It's just that I didn't expect it, based on my understanding of this documentation:
http://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/JavaScript%20Notebook%20Extensions.html

I am running Jupyter core 4.4.0

**Comment by takluyver** [comment]
Yup, I guess that example might be a bit out of date, or oversimplified.

This is what 'trusted' means - it can load potentially dangerous output (like Javascript) without the user having to run any cells. An untrusted notebook will not be able to do that.

**Comment by sjdv1982** [comment]
So I can rely on this feature, good to hear!

But I suppose that there are some cases where you want to have a %%javascript cell for manual triggering, like here:?
https://github.com/jupyter/notebook/issues/1282#issuecomment-318354077

**Comment by takluyver** [comment]
If you want the user to have to run a cell manually, you can clear the output before saving. Normally, though, when you send a notebook to someone else, it won't be trusted on their system until they have run it.

**Comment by sjdv1982** [comment]
> If you want the user to have to run a cell manually, you can clear the output before saving

OK, so the policy is:  a %%javascript cell gets executed unconditionally upon loading/reloading, except for when the output cell is missing in the notebook? 
And the same for %%html, I suppose?

**Comment by sjdv1982** [comment]
> Normally, though, when you send a notebook to someone else, it won't be trusted on their system until they have run it.

That's OK. I am experimenting with notebooks as lightweight web servers. The kernel would run semi-permanently, on my machine.

**Comment by takluyver** [comment]
From Jupyter's point of view, nothing is being executed when you load. The Javascript/HTML code is the output of the cell, and that's saved in the notebook file. When you load it, the output is 'displayed' on the page without going through the kernel. And displaying Javascript causes the browser to run it.

Apart from the trust check, Jupyter doesn't do anything special with Javascript. It's just an output format that makes things happen when it's displayed. And all a `%%javascript` cell does is send its input back as output with the metadata to say it's Javascript.

**Comment by sjdv1982** [comment]
Thanks for the explanation!  

Let me see if I get it right.

%%javascript and %%html, *in trusted mode*, are simply injected directly into the DOM when the notebook is loaded. So they are essentially **data cells** (like markdown or output cells). 
However, if they are in untrusted mode, they behave like **code cells** (even though they don't go through the kernel, of course), that have to be explicitly executed. The underlying "execution" mechanism  is again just a DOM injection, there is no such thing as Jupyter.eval.

Normally, executing a cell puts into trusted mode afterwards (because trusted or untrusted is a feature of the *output* of a cell). 

So, to make a JS or HTML cell that has a code-cell-like behavior (`%%javascript --code` or whatever the syntax ), *it has to remain in a permanently untrusted mode*.

Do I understand it correctly?

**Comment by takluyver** [comment]
Not quite.

`%%javascript` and `%%html` *do* go through the kernel. They're regular code cells, it's just like calling `display(HTML('<span>...'))`. Jupyter does not do anything special with these cells (except for adjusting the syntax highlighting). The kernel sends back the contents as output with the appropriate mimetype so the browser knows what to do with it.

Jupyter handles this output just like it would text or images: it puts it in the page, at which point Javascript runs. Like any other output, it's saved in the notebook file. You can open a notebook file in a text editor and see the Javascript or HTML output saved along with the cell that made it.

When you load a trusted notebook (trust affects the outputs, but it's a property of a whole notebook), the output saved in the file is put back on the page. This causes the Javascript to run again, without involving the kernel, because the Javascript is coming in from the file.

When you load an untrusted notebook, it gets to a Javascript/HTML output and skips putting it on the page, so that that code can't run.

**Comment by takluyver** [comment]
The kernel can send the browser Javascript code as the output of any cell. The model assumes that the browser trusts the kernel.

**Comment by sjdv1982** [comment]
Yes, you are right,  obviously. I have to think of this further.  

What I would like is to use Jupyter notebook/hub/lab eventually as a web portal,  connecting the browser to dockerized kernels. The kernel docker container will be notebook-specific (e.g IPython+Tensorflow) and potentially an external web service. I will have to think of a way to prevent untrusted kernels to access the Jupyter REST API (maybe a sandbox iframe?).

Thank you for your patient explanations. Jupyter is a great project!

**Comment by takluyver** [comment]
If the browser is talking to a notebook server, I don't think you can stop the user playing with the REST API. They don't need to use the kernel for that - they can open the browser's Javascript console and run code in the webpage.

There are a couple of approaches you could take:

- Run the notebook server in the docker container as well, and let the user do what they like with it. This is what Jupyterhub and Binder do - the user only has access to their own notebook server, so they can't interfere with anyone else.
- Run kernels without a notebook server using something like [kernel_gateway](https://github.com/jupyter/kernel_gateway/), and hook them up to web pages that don't try to load/save .ipynb files, e.g. [Thebe](https://github.com/oreillymedia/thebe) (though it looks like Thebe is unmaintained now, but a similar idea is still possible).

**Comment by sjdv1982** [comment]
Thanks!  I don't mind if the user plays with the REST API. I am just worried about untrusted kernels: Docker will protect *my* system, but with the REST API, they could sniff or infect the user's files in Jupyter.

kernel_gateway looks like what I will need. I know next to nothing about browser security, but I suppose that something like Google Caja could sanitize or sandbox the kernel responses.

**Comment by takluyver** [comment]
Yeah, you'd probably need to disable Javascript output entirely, and use something like Caja on HTML, Markdown and SVG output.
